lib/wrap-request.js

```diff
@@ -0,0 +1,59 @@
+/* eslint padded-blocks: 0 */
+module.exports = wrapRequest
+
+const noop = () => Promise.resolve()
+
+function wrapRequest (state, emitter, request, options, retries = 0) {
+  const retryRequest = function (retryAfter) {
+    return new Promise(resolve => setTimeout(resolve, retryAfter * 1000))
+      .then(() => wrapRequest(state, emitter, request, options, retries + 1))
+  }
+
+  const jobOptions = retries > 0 ? { weight: 0, priority: 0 } : {}
+  return state.globalLimiter.schedule(jobOptions, async function () {
+
+    // Guarantee at least 1000ms between writes
+    if (options.method !== 'GET' && options.method !== 'HEAD') {
+      await state.writeLimiter.schedule(noop)
+    }
+
+    // Guarantee at least 5000ms between requests that trigger notifications
+    if (state.triggersNotification[options.url]) {
+      await state.triggersNotificationLimiter.schedule(noop)
+    }
+
+    try {
+      // Execute request
+      return await request(options)
+    } catch (error) {
+      if (error.status === 403 && error.message.toLowerCase().includes('abuse')) {
+        // The user has hit the abuse rate limit.
+        // https://developer.github.com/v3/#abuse-rate-limits
+
+        // Due to known issues with replication lag, the Retry-After header can sometimes be blank
+        // for a few seconds after entering 'abuse limited mode'.
+        // We ensure the Retry-After is always at least 5 seconds.
+        const retryAfter = Math.max(~~error.headers['retry-after'], state.minimumAbuseRetryAfter)
+        emitter.trigger('abuse-limit', retryAfter)
+
+        if (retries < state.retries && state.retries > 0) {
+          return retryRequest(retryAfter)
+        }
+
+      } else if (error.status === 403 && error.headers['x-ratelimit-remaining'] === '0') {
+        // The user has used all their allowed calls for the current time period
+        // https://developer.github.com/v3/#rate-limiting
+
+        const rateLimitReset = new Date(~~error.headers['x-ratelimit-reset'] * 1000).getTime()
+        const retryAfter = Math.max(Math.ceil((rateLimitReset - Date.now()) / 1000), 0)
+        emitter.trigger('rate-limit', retryAfter)
+
+        if (retries < state.retries && state.retries > 0) {
```

Comment:
Do we need to check for `state.retries > 0` here? And if we do, I think I’d switch the order for better readability.

```suggestion
        if (state.retries > 0 && retries < state.retries) {
```

Reply:
The retry stuff is a work in progress. This is all going into a separate plugin today and it will look very different in there.